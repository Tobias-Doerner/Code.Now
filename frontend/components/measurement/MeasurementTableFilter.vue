<template>
  <v-card color="tertiary" flat tile>
    <v-card-actions>
      <v-container fluid>
        <v-row no-gutters>
          <v-col cols="12" sm="12" md="12" lg="6" xl="4">
            <v-combobox
              v-model="selectedCountry"
              :items="countries"
              :label="$t('measurement.label.country')"
              prepend-inner-icon="mdi-filter"
              class="mr-2 my-1"
              hide-details
              item-text="name"
              item-value="code"
              clearable
              dense
              outlined
              @change="handleCountryChange"
            >
              <template v-slot:item="{ item }">
                <div>{{ item.name }}</div>
                <v-spacer />
                <v-chip small>{{ item.code }}</v-chip>
              </template>
            </v-combobox>
          </v-col>

          <v-col cols="12" sm="12" md="12" lg="6" xl="4">
            <v-combobox
              v-model="selectedCity"
              :disabled="selectedCountry === null"
              :items="cities"
              :label="$t('measurement.label.city')"
              prepend-inner-icon="mdi-filter"
              class="mr-2 my-1"
              hide-details
              item-text="name"
              item-value="city"
              clearable
              dense
              outlined
              @change="handleCityChange"
            />
          </v-col>
        </v-row>
      </v-container>
    </v-card-actions>
  </v-card>
</template>

<script>
export default {
  data() {
    return {
      selectedCity: null,
      selectedCountry: null
    }
  },
  computed: {
    cities() {
      if (this.selectedCountry !== null) {
        return this.$store.getters['measurements/getCities'](
          this.selectedCountry.code
        )
      } else {
        return []
      }
    },
    countries() {
      return this.$store.getters['measurements/getCountries']
    }
  },
  methods: {
    emitFilterChange() {
      const el = {}
      el.city = this.selectedCity ? this.selectedCity.city : null
      el.country = this.selectedCountry ? this.selectedCountry.code : null
      this.$emit('filter:changed', JSON.parse(JSON.stringify(el)))
    },
    handleCityChange() {
      this.emitFilterChange()
    },
    handleCountryChange() {
      if (this.selectedCountry === null) {
        this.selectedCity = null
      }
      this.emitFilterChange()
    }
  }
}
</script>
