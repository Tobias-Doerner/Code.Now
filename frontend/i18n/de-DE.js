export default {
  appbar: {
    title: 'AirPoll App'
  },
  common: {
    home: 'Code.Now'
  },
  footer: {
    title: 'Accenture Code.Now Challenge'
  },
  formatter: {
    datetime: 'DD.MM.YY HH:mm:ss'
  },
  page: {
    error: {
      link: 'Zur√ºck zum Start',
      'other error': 'Ein Fehler ist aufgetreten!',
      'page not found': 'Seite nicht gefunden!'
    }
  },
  settings: {
    darkmode: 'Nachtmodus',
    language: 'Sprache',
    title: 'Einstellungen'
  }
}
