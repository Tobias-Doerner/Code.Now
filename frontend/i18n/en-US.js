export default {
  appbar: {
    title: 'AirPoll App'
  },
  common: {
    air_pollution: 'Air Pollution',
    error: {
      network_error: 'Open AQ Platform API is currently not responsive!'
    },
    home: '<Code.Now!>'
  },
  footer: {
    title: 'Accenture Code.Now Challenge'
  },
  formatter: {
    datetime: 'D/M/YY h:mm:ss a'
  },
  measurement: {
    label: {
      city: 'City',
      country: 'Country'
    },
    table: {
      header: {
        air_pollution: 'Air Pollution',
        city: 'City',
        country: 'Country',
        date: 'Date & Time',
        latitude: 'Latitude',
        location: 'Measuring Station',
        longitude: 'Longitude',
        measurement_unit: 'Measurement Unit',
        measurement_value: 'Particulate Matter'
      },
      title: 'Measurements of pollutants'
    }
  },
  page: {
    error: {
      link: 'Back to start',
      'other error': 'An error occurred!',
      'page not found': 'Page not found!'
    },
    index: {
      card: {
        btn: 'Go to the Source',
        text:
          'Front-end of the AirPoll App implemented with Vue.js, Nuxt.js and Vuetify.'
      }
    }
  },
  settings: {
    darkmode: 'Dark Mode',
    language: 'Language',
    title: 'Settings'
  }
}
