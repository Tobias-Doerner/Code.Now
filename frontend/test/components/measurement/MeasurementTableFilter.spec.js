import Vue from 'vue'
import Vuetify from 'vuetify'
import Vuex from 'vuex'
import { createLocalVue, mount, shallowMount } from '@vue/test-utils'
import { setupI18n } from '@/test/i18n'
import { setupVuetify } from '@/test/vuetify'

import MeasurementTableFilter from '@/components/measurement/MeasurementTableFilter'

Vue.use(Vuetify)
const localVue = createLocalVue()
localVue.use(Vuex)
const i18n = setupI18n(localVue)
const vuetify = setupVuetify(localVue)

describe('MeasurementTableFilter.vue', () => {
  const state = {}
  const getters = {
    'measurements/getCities': () => (code) => {
      if (code === 'CH') {
        return [
          { city: 'Bern', name: 'Bern' },
          { city: 'Genève', name: 'Genève' },
          { city: 'Lucerne', name: 'Lucerne' }
        ]
      } else {
        return []
      }
    },
    'measurements/getCountries': () => {
      return [
        { code: 'BE', name: 'Belgium' },
        { code: 'BG', name: 'Bulgaria' },
        { code: 'BH', name: 'Bahrain' }
      ]
    }
  }
  const mutations = {}
  const actions = {}
  const store = new Vuex.Store({ state, actions, getters, mutations })

  const cmp = shallowMount(MeasurementTableFilter, {
    localVue,
    store,
    i18n,
    vuetify
  })

  describe('computed', () => {
    describe('cities', () => {
      it('should return array of cities', () => {
        cmp.vm.selectedCountry = {
          code: 'CH',
          name: 'Switzerland'
        }
        expect(cmp.vm.cities).toEqual([
          { city: 'Bern', name: 'Bern' },
          { city: 'Genève', name: 'Genève' },
          { city: 'Lucerne', name: 'Lucerne' }
        ])
      })

      it('should return empty array of cities', () => {
        cmp.vm.selectedCountry = {
          code: 'DE',
          name: 'Germany'
        }
        expect(cmp.vm.cities).toEqual([])
      })
    })

    describe('countries', () => {
      it('should return array of countries', () => {
        expect(cmp.vm.countries).toEqual([
          { code: 'BE', name: 'Belgium' },
          { code: 'BG', name: 'Bulgaria' },
          { code: 'BH', name: 'Bahrain' }
        ])
      })
    })
  })

  describe('methods', () => {
    describe('emitFilterChange', () => {
      it('should emit filter:changed event with payload', async () => {
        cmp.vm.selectedCity = null
        cmp.vm.selectedCountry = {
          code: 'CH',
          name: 'Switzerland'
        }
        cmp.vm.emitFilterChange()

        await cmp.vm.$nextTick()

        expect(cmp.emitted('filter:changed')[0]).toEqual([
          { city: null, country: 'CH' }
        ])
      })

      it('should emit filter:changed event with payload', async () => {
        cmp.vm.selectedCity = { city: 'Lucerne', name: 'Lucerne' }
        cmp.vm.selectedCountry = null
        cmp.vm.emitFilterChange()

        await cmp.vm.$nextTick()

        expect(cmp.emitted('filter:changed')[1]).toEqual([
          { city: 'Lucerne', country: null }
        ])
      })

      it('should emit filter:changed event with payload', async () => {
        cmp.vm.selectedCity = { city: 'Lucerne', name: 'Lucerne' }
        cmp.vm.selectedCountry = {
          code: 'CH',
          name: 'Switzerland'
        }
        cmp.vm.emitFilterChange()

        await cmp.vm.$nextTick()

        expect(cmp.emitted('filter:changed')[2]).toEqual([
          { city: 'Lucerne', country: 'CH' }
        ])
      })

      it('should emit filter:changed event with payload', async () => {
        cmp.vm.selectedCity = null
        cmp.vm.selectedCountry = null
        cmp.vm.emitFilterChange()

        await cmp.vm.$nextTick()

        expect(cmp.emitted('filter:changed')[3]).toEqual([
          { city: null, country: null }
        ])
      })
    })

    describe('handleCityChange', () => {
      it('should call emitFilterChange method', async () => {
        cmp.vm.selectedCity = null
        cmp.vm.selectedCountry = null
        cmp.vm.handleCityChange()

        await cmp.vm.$nextTick()

        expect(cmp.emitted('filter:changed')[4]).toEqual([
          { city: null, country: null }
        ])
      })
    })

    describe('handleCountryChange', () => {
      it('should call handleCountryChange method', async () => {
        cmp.vm.selectedCity = { city: 'Lucerne', name: 'Lucerne' }
        cmp.vm.selectedCountry = {
          code: 'CH',
          name: 'Switzerland'
        }
        cmp.vm.handleCountryChange()

        await cmp.vm.$nextTick()

        expect(cmp.emitted('filter:changed')[5]).toEqual([
          { city: 'Lucerne', country: 'CH' }
        ])
      })

      it('should call handleCountryChange method', async () => {
        cmp.vm.selectedCity = { city: 'Lucerne', name: 'Lucerne' }
        cmp.vm.selectedCountry = null
        cmp.vm.handleCountryChange()

        await cmp.vm.$nextTick()

        expect(cmp.emitted('filter:changed')[6]).toEqual([
          { city: null, country: null }
        ])
      })
    })
  })

  describe('snapshot', () => {
    it('should match snapshot', () => {
      const wrapper = mount(MeasurementTableFilter, {
        localVue,
        store,
        i18n,
        vuetify
      })

      expect(wrapper.element).toMatchSnapshot()
    })
  })
})
