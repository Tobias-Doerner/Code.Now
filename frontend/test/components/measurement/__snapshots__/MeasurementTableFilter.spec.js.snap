// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MeasurementTableFilter.vue snapshot should match snapshot 1`] = `
<div
  class="v-card v-card--flat v-sheet theme--light rounded-0 tertiary"
>
  <div
    class="v-card__actions"
  >
    <div
      class="container container--fluid"
    >
      <div
        class="row no-gutters"
      >
        <div
          class="col-sm-12 col-md-12 col-lg-6 col-xl-4 col-12"
        >
          <div
            class="v-input mr-2 my-1 v-input--hide-details v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined v-select v-autocomplete"
          >
            <div
              class="v-input__control"
            >
              <div
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-owns="list-11"
                class="v-input__slot"
                role="combobox"
              >
                <div
                  class="v-input__prepend-inner"
                >
                  <div
                    class="v-input__icon v-input__icon--prepend-inner"
                  >
                    <i
                      aria-hidden="true"
                      class="v-icon notranslate mdi mdi-filter theme--light"
                    />
                  </div>
                </div>
                <fieldset
                  aria-hidden="true"
                >
                  <legend
                    style="width: 0px;"
                  >
                    <span>
                      ​
                    </span>
                  </legend>
                </fieldset>
                <div
                  class="v-select__slot"
                >
                  <label
                    class="v-label theme--light"
                    for="input-11"
                    style="left: 0px; position: absolute;"
                  >
                    Land
                  </label>
                  <input
                    autocomplete="off"
                    id="input-11"
                    type="text"
                  />
                  <div
                    class="v-input__append-inner"
                  >
                    <div
                      class="v-input__icon v-input__icon--clear"
                    >
                      <button
                        aria-label="clear icon"
                        class="v-icon notranslate v-icon--disabled v-icon--link mdi mdi-close theme--light"
                        disabled="disabled"
                        type="button"
                      />
                    </div>
                  </div>
                  <div
                    class="v-input__append-inner"
                  >
                    <div
                      class="v-input__icon v-input__icon--append"
                    >
                      <i
                        aria-hidden="true"
                        class="v-icon notranslate mdi mdi-menu-down theme--light"
                      />
                    </div>
                  </div>
                  <input
                    type="hidden"
                  />
                </div>
                <div
                  class="v-menu"
                >
                  <!---->
                </div>
              </div>
            </div>
          </div>
        </div>
         
        <div
          class="col-sm-12 col-md-12 col-lg-6 col-xl-4 col-12"
        >
          <div
            class="v-input mr-2 my-1 v-input--hide-details v-input--is-disabled v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined v-select v-autocomplete"
          >
            <div
              class="v-input__control"
            >
              <div
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-owns="list-16"
                class="v-input__slot"
                role="combobox"
              >
                <div
                  class="v-input__prepend-inner"
                >
                  <div
                    class="v-input__icon v-input__icon--prepend-inner"
                  >
                    <i
                      aria-hidden="true"
                      class="v-icon notranslate v-icon--disabled mdi mdi-filter theme--light"
                    />
                  </div>
                </div>
                <fieldset
                  aria-hidden="true"
                >
                  <legend
                    style="width: 0px;"
                  >
                    <span>
                      ​
                    </span>
                  </legend>
                </fieldset>
                <div
                  class="v-select__slot"
                >
                  <label
                    class="v-label v-label--is-disabled theme--light"
                    for="input-16"
                    style="left: 0px; position: absolute;"
                  >
                    Stadt
                  </label>
                  <input
                    autocomplete="off"
                    disabled="disabled"
                    id="input-16"
                    type="text"
                  />
                  <div
                    class="v-input__append-inner"
                  >
                    <div
                      class="v-input__icon v-input__icon--clear"
                    >
                      <button
                        aria-label="clear icon"
                        class="v-icon notranslate v-icon--disabled v-icon--link mdi mdi-close theme--light"
                        disabled="disabled"
                        type="button"
                      />
                    </div>
                  </div>
                  <div
                    class="v-input__append-inner"
                  >
                    <div
                      class="v-input__icon v-input__icon--append"
                    >
                      <i
                        aria-hidden="true"
                        class="v-icon notranslate v-icon--disabled mdi mdi-menu-down theme--light"
                      />
                    </div>
                  </div>
                  <input
                    type="hidden"
                  />
                </div>
                <div
                  class="v-menu"
                >
                  <!---->
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
